src/Control/Effect/Cut.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleContexts, FlexibleInstances, LambdaCase, MultiParamTypeClasses, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Cut
+( Cut(..)
+, cutfail
+, call
+, cut
+, Branch(..)
+, branch
+, runCut
+, CutC(..)
+) where
+
+import Control.Applicative (Alternative(..))
+import Control.Effect.Carrier
+import Control.Effect.Internal
+import Control.Effect.NonDet
+import Control.Effect.Sum
+
+-- | 'Cut' effects are used with 'NonDet' to provide control over backtracking.
+data Cut m k
```

Comment:
As discussed with @aymannadeem, a `Cut` effect is described in _Effect Handlers in Scope_ §5, as an example of a non-orthogonal effect: i.e. an effect that interacts with other effects (in this case, nondeterminism).